Parsing THF and adding several regressions

HanielB opened the pull request:
Modifies the TPTP grammar to allow parsing THF, its higher-order extension. Also adds several regressions that were instrumental for exercising features of the parser and of higher-order solving in CVC4. When possible the actual exercised feature was marked in the regression name.

ajreynol commented on src/options/options.h:
```diff
@@ -241,7 +241,7 @@ class CVC4_PUBLIC Options {
   unsigned getThreadStackSize() const;
   unsigned getThreads() const;
-
+  bool getHoFlattenTotal() const;
```
I would prefer if this was not a parsing option. Can you instead implement this as a rewrite internally after parsing?

HanielB replied:
I'll purge it from this PR and add in another one then. This option is actually not really impactful, although sometimes it helps

ajreynol commented on src/parser/tptp/Tptp.g:
```diff
@@ -637,8 +759,320 @@ fofBinaryNonAssoc[CVC4::parser::tptp::NonAssoc& na]
+      // flatten body type
+      Type range = expr.getType();
+      std::vector<Expr> flattenVars;
+      if (range.isFunction())
```
There is a utility that does this block of code (almost exactly I think): https://github.com/CVC4/CVC4/blob/master/src/parser/parser.h#L592.
Would it be easy just to call this function?

HanielB replied:
That was actually my inspiration. IIRC though, that is used only for types, while here it's about creating the term. While the loop part is the same between the two pieces of code, what they actually produce is different.

HanielB replied:
I changed the code to use the existing function. There will be the spurious computation of some type information, but that is negligible.

ajreynol commented on src/parser/tptp/Tptp.g:
```diff
@@ -637,8 +759,303 @@ fofBinaryNonAssoc[CVC4::parser::tptp::NonAssoc& na]
+    )?
+  ;
+
+// atom
```
What are these comments?

ajreynol commented on src/parser/tptp/Tptp.g:
```diff
@@ -107,6 +107,7 @@ using namespace CVC4::parser;
 #include "parser/antlr_input.h"
 #include "parser/parser.h"
 #include "parser/tptp/tptp.h"
+#include "options/options.h"
```
Do we still need this?

ajreynol commented:
Looks good to me, although there is a failing regression now?

ajreynol commented:
Failing regression is `regress0/ho/nested_lambdas-AGT034^2.smt2`, looks to be a timeout.

HanielB commented:
Issue was with proof checking. Fixed now (by disabling it for that regression).

HanielB commented:
I updated the regressions to respect these criteria (I was checking the times only on production). I also moved a bunch of the old ho regressions (in fact, all but one) from regress0 to regress1 because they failed the criteria.
